src/radical/pilot/agent/launch_method/prte2.py

```diff
@@ -82,141 +95,195 @@ def rm_config_hook(cls, name, cfg, rm, log, profiler):
-            prte += ' --mca orte_state_base_verbose 1'
-
-        # large tasks imply large message sizes and we need to account for that
-        _base_max_msg_size = 1024 * 1024 * 1024 * 1
-        # FIXME: we should derive the message size from DVM size - smaller DVMs
-        #        will never need large messages, as they can't run large tasks
-
-        # version 1 vs 2
-        # prte += ' --pmca ptl_base_max_msg_size %d' % _base_max_msg_size
-        # # prte += ' --pmca rmaps_base_verbose 5'
-        prte += ' --mca ptl_base_max_msg_size %d' % _base_max_msg_size
-
-        # debug mapper problems for large tasks
-        if log.isEnabledFor(logging.DEBUG):
-            # version 1 vs 2
-            # prte += ' -pmca orte_rmaps_base_verbose 100'
-            prte += ' --mca orte_rmaps_base_verbose 100'
-
-        # 2 temporary tweaks on Summit (not needed for the long run)
-        # (1) avoid 64 node limit (ssh connection limit)
-        # prte += ' --pmca plm_rsh_no_tree_spawn 1'
-        # (2) ensure 1 ssh per dvm
-        # prte += ' --pmca plm_rsh_num_concurrent %d' % vm_size
-
-        # version 2
-        prte += ' --mca plm_rsh_no_tree_spawn 1'
-        prte += ' --mca plm_rsh_num_concurrent %d' % vm_size
-        # FIXME: these 2 options above might not be needed
-        #        (corresponding experiments should be conducted)
-
-        # Use (g)stdbuf to disable buffering.  We need this to get the
-        # "DVM ready" message to ensure DVM startup completion
-        #
-        # The command seems to be generally available on Cray's,
-        # if not, we can code some home-cooked pty stuff (TODO)
-        stdbuf_cmd = ru.which(['stdbuf', 'gstdbuf'])
-        if not stdbuf_cmd:
-            raise Exception('(g)stdbuf not found')
-        stdbuf_arg = '-oL'
-
-        # base command = (g)stdbuf <args> + prte + prte_args + debug_args
-        cmd = '%s %s %s ' % (stdbuf_cmd, stdbuf_arg, prte)
-
-        # additional (debug) arguments to prte
-        verbose = bool(os.environ.get('RADICAL_PILOT_PRUN_VERBOSE'))
-        if verbose:
-            cmd += ' '.join(['--debug-devel',
-                             '--mca odls_base_verbose 100',
-                             '--mca rml_base_verbose 100'])
-
-        cmd = cmd.strip()
-
-        log.info('start prte on %d nodes [%s]', vm_size, cmd)
-        profiler.prof(event='dvm_start', uid=cfg['pid'])
-
-        dvm_process = mp.Popen(cmd.split(), stdout=mp.PIPE, stderr=mp.STDOUT)
+        # get `dvm_count` from resource config
+        # FIXME: another option is to introduce `nodes_per_dvm` in config
+        #        dvm_count = math.ceil(len(rm.node_list) / nodes_per_dvm)
+        dvm_count      = cfg['resource_cfg'].get('dvm_count') or 1
+        nodes_per_dvm  = math.ceil(len(rm.node_list) / dvm_count)
+
+        dvm_uri_list   = []
+        dvm_hosts_list = []
+
+        dvm_ready_list = []  # threading.Event that checks DVM's readiness
```

Comment:
You have three lists which need to have the same synced index schema.  Would it make sense to switch to a dict like
```py
dvm_info = {
    'dvm_id': {
        'uri' ; ...,
        'hosts': [...],
        'readly': mt.Event()
    }
        ...
}
```
?

Reply:
so, I thought about this for couple days :)
1) I would think that `dvm_ready` flags should be separate (from other dvm-related info, which will go to `lm_info`), since it is [temporary] internal data, and we don't modify a corresponding list variable (that keeps them), where index of each flag corresponds to `dvm_id`
2) I agree to keep together `dvm_id`, `uri` and `hosts`, but would the following structure fit better?
```python
# record format: (<uri>, <dvm_id>): [<node_id>,...]
        dvm_info = {
            ('prte-batch464440.0;tcp4://10.134.0.44:56089', 0): ['node.0000', 'node.0001']
        }
.....
        lm_info = {'dvm_info'    : dvm_info,
                   'version_info': prte_info,
                   'cvd_id_mode' : 'physical'}
# thus to have key `dvm_info` instead of previous two keys `dvm_uri` and `dvm_hosts`
```

Reply:
> **edit:** OR dvm record format `<dvm_id>: (<uri>, [<node_id>,...])`

At the moment, the scheduler sets the `dvm_uri` for the task.  Eventually, it should only set a partition ID, and then we need to look up dvm_uri and nodes via that partition ID.  In that case, the structure above works well IMHO.

Reply:
#2323